src/Terminal/ui/TerminalInput.tsx

```diff
@@ -237,6 +265,27 @@ export function TerminalInput(): React.ReactElement {
       if (len == 0) {
         return;
       }
+
+      // If there is a partial command in the terminal, hitting "up" will filter the history
+      if (value && !autofilledValue && Settings.EnableHistorySearch) {
+        if (searchResults.length > 1) {
+          const results = [...searchResults];
+          results.push(results.shift() ?? "");
+          setSearchResults(results);
+          return;
+        }
+        const newResults = [...new Set(Terminal.commandHistory.filter((item) => item?.startsWith(value)).reverse())];
```

Comment:
Yes, to clean up the history. Also it's built out to make a new array like that because React doesn't re-render if a value in the array changes, only if the array itself is updated.